plugin/pkg/admission/podpreset/admission.go

```diff
@@ -255,14 +255,22 @@ func mergeEnv(envVars []api.EnvVar, podPresets []*settingsv1alpha1.PodPreset) ([
 func mergeEnvFrom(envSources []api.EnvFromSource, podPresets []*settingsv1alpha1.PodPreset) ([]api.EnvFromSource, error) {
 	var mergedEnvFrom []api.EnvFromSource
 
+	// merge envFrom using structural identity to ensure Admit reinvocations are idempotent
+	origEnvSources := map[api.EnvFromSource]struct{}{}
+	for _, envSource := range envSources {
+		origEnvSources[envSource] = struct{}{}
```

Comment:
EnvFromSource has pointer fields, so equality checks don't work with round-tripped/copied objects

adding this to `CheckSafeAdmissionReinvocation` exposes that:
```diff
diff --git a/test/utils/admission.go b/test/utils/admission.go
index 4d5b192d1cd..7a33ce6e735 100644
--- a/test/utils/admission.go
+++ b/test/utils/admission.go
@@ -35,6 +35,14 @@ func CheckSafeAdmissionReinvocation(t *testing.T, attr admission.Attributes, tes
                if attr.GetObject() != nil {
                        outputs = append(outputs, attr.GetObject().DeepCopyObject())
                }
+
+               // replace attr.GetObject() with a copy of itself to make sure admission is safe to reinvoke with a round-tripped copy
+               if deepCopyInto, ok := reflect.TypeOf(attr.GetObject()).MethodByName("DeepCopyInto"); ok {
+                       deepCopyInto.Func.Call([]reflect.Value{
+                               reflect.ValueOf(attr.GetObject().DeepCopyObject()),
+                               reflect.ValueOf(attr.GetObject()),
+                       })
+               }
        }
```

Reply:
I was wondering what the implications of pointer fields were for map equality. I'll have to provide identity without using structural identity. Thanks for the diff to `CheckSafeAdmissionReinvocation`, I'll apply that.

Reply:
reviewers: I've moved the code around, the logic from `CheckSafeAdmissionReinvocation is now in `reinvoker.Admit`: https://github.com/kubernetes/kubernetes/pull/78269/commits/fa1601d530eb7d0c92d0a3306810b659ef1db1a6#diff-6c062c69b65be6a145a073152afb116dR42

Reply:
I've applied this and switched to using a map key that should be safe and correct + a fuzz test for it to ensure it stays correct.